File: articles/differential-equations-population-physiological-modeling/c/differential_equations_kernel.c
/*
 * Compact differential-equation biology numerical kernel in C.
 */

#include <math.h>
#include <stdio.h>

double logistic_final(double n0, double r, double k, double dt, double t_end) {
    int steps = (int)floor(t_end / dt) + 1;
    double n = n0;

    for (int i = 1; i < steps; i++) {
        double dn = r * n * (1.0 - n / k);
        n = fmax(n + dn * dt, 0.0);
    }

    return n;
}

double homeostasis_final(double x0, double set_point, double k, double dt, double t_end) {
    int steps = (int)floor(t_end / dt) + 1;
    double x = x0;

    for (int i = 1; i < steps; i++) {
        double dx = -k * (x - set_point);
        x += dx * dt;
    }

    return x;
}

double pk_final(double c0, double elimination_rate, double dt, double t_end) {
    int steps = (int)floor(t_end / dt) + 1;
    double c = c0;

    for (int i = 1; i < steps; i++) {
        double dc = -elimination_rate * c;
        c = fmax(c + dc * dt, 0.0);
    }

    return c;
}

int main(void) {
    printf("logistic_final=%.6f\n", logistic_final(100.0, 0.30, 2000.0, 0.05, 40.0));
    printf("homeostasis_final=%.6f\n", homeostasis_final(180.0, 100.0, 0.18, 0.05, 30.0));
    printf("pk_final=%.6f\n", pk_final(20.0, 0.12, 0.05, 48.0));
    return 0;
}

File: articles/differential-equations-population-physiological-modeling/fortran/differential_equations_kernel.f90
! Compact differential-equation biology numerical kernel in Fortran.

program differential_equations_kernel
  implicit none

  integer :: step, steps
  real :: dt
  real :: N, r, K, dN
  real :: x, set_point, k_home, dx
  real :: C, elimination_rate, dC
  real :: S, I, R, beta, gamma, dS, dI, dR, peak_I, time_to_peak

  dt = 0.05

  N = 100.0
  r = 0.30
  K = 2000.0
  steps = int(40.0 / dt) + 1

  do step = 2, steps
    dN = r * N * (1.0 - N / K)
    N = max(N + dN * dt, 0.0)
  end do

  x = 180.0
  set_point = 100.0
  k_home = 0.18
  steps = int(30.0 / dt) + 1

  do step = 2, steps
    dx = -k_home * (x - set_point)
    x = x + dx * dt
  end do

  C = 20.0
  elimination_rate = 0.12
  steps = int(48.0 / dt) + 1

  do step = 2, steps
    dC = -elimination_rate * C
    C = max(C + dC * dt, 0.0)
  end do

  S = 0.99
  I = 0.01
  R = 0.0
  beta = 0.35
  gamma = 0.10
  peak_I = I
  time_to_peak = 0.0
  steps = int(120.0 / dt) + 1

  do step = 2, steps
    if (I > peak_I) then
      peak_I = I
      time_to_peak = real(step - 1) * dt
    end if

    dS = -beta * S * I
    dI = beta * S * I - gamma * I
    dR = gamma * I

    S = max(S + dS * dt, 0.0)
    I = max(I + dI * dt, 0.0)
    R = max(R + dR * dt, 0.0)
  end do

  print *, "Logistic final:", N
  print *, "Homeostasis final:", x
  print *, "PK final:", C
  print *, "SIR peak infected:", peak_I
  print *, "SIR time to peak:", time_to_peak
  print *, "SIR final recovered:", R
end program differential_equations_kernel
